### ISSUE TITLE
Disconnection Glitch

### ISSUE THREAD

**Issue by machiavel23** [Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab#150]
Ok, I'm really having a hard time with this one, especially because there is really few real feedback on it except "muh codejerk's shit won't run reeeeeeeee"

Obviously, it's a bit hard to find and fix an issue if you don't have many info on it. So I open this issue on github in hope to get some help.

You all know the disconnection problem occurring since the Brainlets Edition of the colab (aka one-cell update). I definitely don't manage to resolve it, and apparently won't be able to.

According to my latest tests, the code itself works correctly (excepts that thadunge repo seems unusable on this colab since his recent changes, but I'll check that later. clover works fine, so we'll do the test on his version) but apparently there's either some kind of idling popping up on the virtual machine causings the disconnection, or a too short keepalive of the google login. I have only found 2 workarounds so far but haven't had time to check them myself yet:

1/ Click on the RAM/DISK button from time to time to prevent the VM from going idle

2/ using a mini function to paste in the browser's console that will emulate a click each 60 seconds:

function ClickConnect(){
console.log("Working");
document.querySelector("colab-toolbar-button#connect").click()
}
setInterval(ClickConnect,60000)

If you know ANYTHING else that could help or bypass solve the bug, or at least confirm if it is indeed the reason of the bug, please help.

any work around this should be made from the beta-ocyoa-1 branch as much as possible.

Thanks.

**Comment by machiavel23** [comment]
So I did a few more tests, and it seems retards STILL can't read. Lemme explain my latest results (beta colab, clover edition)

playing from the github link -> disconnected after 10 minutes

copying to gdrive and playing from there -> more than 40 minutes before i got bored and stopped it myself.

Maybe there's NO disconnect glitch BUT retards who jusat can't fucking READ and fucking copy the fucking file in their fucking GDrive.

And since I barely have any fucking results when I ask those fucking coomerbrains to do a fucking simple test, I will implement an url detection at the very beginning than will make sure the cell will stop if it detect "github" in the page url instead of "drive" with BIG message in the ouput.

If this REALLY solves the problems, then thanks retards, give me back my 4 fucking days of life I lost trying to figure out an inexistant bug.

**Comment by machiavel23** [comment]
Edit: link remove

gonna wait for kornilov to check the pull request, and if he merge, i will rebranch beta from there.

**Comment by machiavel23** [comment]
acually, I re open to make sure you see it. Once done you can close

**Comment by VBPXKSMI** [comment]
Saw it, also commented on the thread, if it really is a retards falsepositive then we can move on.
I have just one question, does the beta Colab have improved code? As in, is it more fleshed out than the one in the main branch?

**Comment by machiavel23** [comment]
nope. That's why I'm gonna close this one and rebranch from the lastest version. The few modifications that were made in the actual beta are mostly cosmetics and can easily be redone quickly.

**Comment by VBPXKSMI** [comment]
I'm afraid I confirmed that the disconnect issue "seems" real, got it myself.
I'm in the process of testing some code to solve the issue (don't even ask what or how), will come back after testing that no disconnect happens. (but don't expect what I'm doing to work, so get high hopes)

**Comment by machiavel23** [comment]
apparently, putting a line of code after the play.py works. damn wizardry.... anyway, i'm deleting the actual beta, and will rebranch from 1.21. leaving the issue open until we're sure it doesn't disconnect.

I took the time to think about it, and I have only two guesses about the "why".

1/ typing on the keyboard to play the game doesn't make colab react to an actual user input. Since apparently it's made to detect clicks, it would make sense that it goes idle quickly if there's no mouse move. And that's why putting a line of code after play.py prevents this. Considering the line being a part of the cell code, the cell is considered as "non finished running", and so, it prevents the disconnection. One question still remains however with that theory: Why since the merge and not before? It should have disconnect since the beginning. Or maybe the hype around it made people frequently switch from the game to the board, hence preventing the colab to disconnect back then... yeah, maybe...

2/ The time. I've noticed much more activity in the thread when it was around normal hours for americans, and so, much more disconnections reports. I haven't been able to say if there was proportionnally more disconnections however, I'm not a statistician. maybe there is just a big flow of people trying to access it more or less at the same time and colab's servers are panicking? then again, why now and not before, when the thread was quickly reaching bump limit?

**Comment by machiavel23** [comment]
even running it straight from the github link do not disconnect. i have a game runnning since an hour now without even being on the tab.  guess it's settled

**Comment by VBPXKSMI** [comment]
Fucking damn wizardry.
I will make a post asking for feedback to make sure, just in case.

As for the reason why this works I'm still puzzled.
1/ Could be a possible theory, if it wasn't like you stated that this wasn't happening before, although is true that maybe the hype made people don't notice it. But then, why isn't anyone reporting issues with the v1.0 right now? I don't think this theory is correct, at least not entirely.
2/ This one may be more plausible, but I'm still not sure, it sounds too convenient of a reason to be real, yeah more people means more work for the servers but then, why didn't we see such problems when Reddit discovered our Colab? Again, I'm not sure this is the answer.
3/ My theory is that, when everything is concentrated into just 1 cell, the Colab cannot handle it and it randomly disconnects, this may be related to what you said in point one about not taking inputs unless clicking. Why? I don't know. But maybe now that there is a code that leaves the cell running in the background the Colab can, somehow, handle it. But again, I'm not sure.

We need tests to be done about if they current code line actually improves things by itself or if any code can get the work done (you kinda proved this but I will ask feedback from Anons in the thread).

**Comment by machiavel23** [comment]
best way to prove if it any line does the trick is to discretly replace it with a print('something'). don't make an announce and don't change the filename, they won't notice it.

**Comment by VBPXKSMI** [comment]
Ok will do that, after a day or two.
But I still want to be sure if it makes the responses slower or not, so we need feedback for that, no?

Also, will putting the missing pip dependencies do the trick or do you want me to wait for you to clean the code properly?

**Comment by machiavel23** [comment]
don't worry about the pip, i've handle it in the beta. the pip lines are removed and replaced with a single

!pip intall -r requirements.txt

it will only install the dependencies needed by the repo selected. placing this line just between the !git clone and the !./ install.sh for the initial setup, and placing it just before the git fetch for the replay will optimize it all.

I  did noticed more latency since the new line was added after !python play.py when it was your ariable, but it seems the trick still works and and doesn't seems to cause latency with a print, at least with my tests. a few feed backs with the print should be enough to confirm this.

**Comment by VBPXKSMI** [comment]
One Anon reported to not have seen any kind of difference in response time with both code lines.
What he did point out was that when he used the "print..." code, he felt that there was a faster "generating or initiallising story", but this may just be anecdotal so we need more feedback.

**Comment by uotalkie** [comment]
Hello,
I just sign up to leave this message because I live outside US and cannot write in 4chan's board.

Assuming long code in all-in-one is the main problem of disconnection issue,
I have an idea to solve the issue.

The idea is as follows:
* separate install part and play part like old version 1.0
* make a new "all-in-one" cell to execute install part and play part

Here is my solution. Because of lack of skill about python, I disabled clover edition on this notebook.
I don't know how to pass arguments between code cells.

[Untitled0.zip](https://github.com/Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab/files/4003276/Untitled0.zip)

- or find it here:
https://drive.google.com/open?id=1nSWzppO8G5-QSTSZHdzYbNsobTBmQD-1 

I hope this helps resolving the disconnection issue.

**Comment by VBPXKSMI** [comment]
While making an all in one cell for both the install and the play part would be way better than what we had in v1.0, I still feel we should strive to make the current one in v1.21 work.
The reason is simple, is the most straightforward, direct and easy to use Colab we will get for a long while. We made this to help brainlets, and to automatize things as much as possible.

The disconnection bug seems mostly fixed in v1.21.2, yes we still need feedback to see how we can improve it, but it works. Recently some Anons have reported to still have the issue, but since they didn't specify how or what they were playing and in what circumstances, we don't have enough info to conclude if the bug still exist.
Until further confirmation, I consider the bug solved for the most part.

Still, thank you for your support. I will let machiavel23 (visualfag) review what you did, and let him consider if it's worth doing anything with it or not.

**Comment by machiavel23** [comment]
I actually find some valuable information in this notebook, it could gt me an interesting idea. @uotalkie if you could provide me the reference you used, it would be a great help actually

**Comment by uotalkie** [comment]
I have no idea what "reference" you are saying, but I will try to explain.

First of all, I had a strong feeling that long codes of the notebook causes disconnection,
so I tried to shorten the codes as short as possible.

I tried to copy the all-in-one code into a .py file and call from the notebook, but I failed
because ipython did not execute !command in .py file.

I also tried .sh with #!ipython but it didn't work either.

Then an idea came that calling a cell from another cell, and found execute_cell(s) method,
but google colab didn't seem to work. Here's the code about it:
https://stackoverflow.com/questions/47567834/execute-a-jupyter-notebook-cell-programmatically

Lastly I found this solution, not exactly call a cell from another cell, but bring codes
from a notebook file and execute them. Here's the code I brought to my test notebook:
https://stackoverflow.com/questions/22328052/ipython-notebook-programmatically-read-and-execute-cells

That's all I can tell. I know my codes need many improvements, but I can't help anymore
because I am not familiar to python - I even had to search the net to get the usage
of if and for statements in python.

**Comment by machiavel23** [comment]
yeah, I had this python level too a week ago... Anyway, I looked at your links. To do something neat and tidy, it would need to install the nbextensions in the notebook. I already thought about that, but I'm not sure if people would have them after copying the colab. Thanks for the links however, it's always useful to see that kind of tricks

**Comment by VBPXKSMI** [comment]
>if people would have them after copying the colab

We should do a small test and see if that's the case.

**Comment by uotalkie** [comment]
Hi,

I have another idea.

Because my previous code is not really calling another cell, but get code from another file
and execute the code, it is possible to separate code notebook and wrapper notebook.

So.. I made a wrapper that bring actual code notebook from github and take and execute
the code from that background notebook.

Here is the wrapper so take a look.
[wrapper.zip](https://github.com/Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab/files/4006991/wrapper.zip)

- or here:
https://drive.google.com/open?id=1nqrkAF8GhO7S-LDZr7fZyGQvfPQlGztQ

Passing arguments was not so hard. Just put them in the variables and it works.
Because of the default values it doesn't work in original Open_CYOAI_Colab.ipynb,
but when I copied the notebook for a test and disabled default values, I can play
both thadunge2 and clover.

I am curious if this wrapping trick solves the issue or not. How about taking a beta test
for original Open_CYOAI_Colab versus wrapper with 4chan gangs?

**Comment by VBPXKSMI** [comment]
Ummm, if this works we could make some tests comparing it to the current Beta code when it's finally merged into the main branch.

Honestly, this disconnect bug has drained all my energy over the past week, hopefully we can start the new year focusing in other QoL changes (I'm making 2 pastebins at the momment, one of them will be directly linked in the Colab).

**Comment by machiavel23** [comment]
yeah, I'd like this glitch to be settled too to finally have feedbacks on the new installation method, I can't go further because of that. It's annoying because my next plan is to restrict the git clone to only master branch and add an access to clover's config.ini. I's like to work but these retards keep staying on olders colabs until the glitch is fiwed.... as if it was feasible without them feedbacking... anyway, I'm quite convinced it's just some random shit happening with google. I've seen people mention havnig no problem a day, and many problems another. I've also seen notebooks with quite lenghty cells (like the one with the storages) and they don't have problems, I doubt the code lenght is at fault here.

@uotalkie Hmm... Your code is an idea indeed, but it'll need to be reajusted. I've tried to run it but it seems the wat you do things is deprecated thus it sends a warning. it does seems to work however. I'll investigate further. got nothing else to do until the retards realize we need their feedback, anyway...

**Comment by uotalkie** [comment]
> I've tried to run it but it seems the wat you do things is deprecated thus it sends a warning.

Oh, that's easy (not easy for me because I had to search the python documents and do trial-and-errors for a couple of hours :P ). Just replacing deprecated methods to latest replaces and it's good to go.

Here's the revised deprecated-error-free version:
[wrapper.zip](https://github.com/Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab/files/4007799/wrapper.zip)

- or here:

https://drive.google.com/open?id=1nqrkAF8GhO7S-LDZr7fZyGQvfPQlGztQ

**Comment by machiavel23** [comment]
yep. it works. I'm currently rebuilding the all in one cell from scratch in another file to test something. meanwhile, if you could redirect your wrapper to https://github.com/Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab/blob/beta/Open_CYOAI_Colab_1.22.3_BETA.ipynb and share it in a default usable state (can't run it by default for now, have to switch mode) it could be useful to check if the dc occurs with it

**Comment by uotalkie** [comment]
I feel my python skill going up ;)

Here's another revised wrapper.
[wrapper.zip](https://github.com/Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab/files/4008044/wrapper.zip)

- or here:
https://drive.google.com/open?id=1nqrkAF8GhO7S-LDZr7fZyGQvfPQlGztQ

In this version,
- added ipynb_file and branch variables to decide which branch and which file to get
- added is_wrapped variable to tell wrappee whether it is wrapped or not
   (read comments in the code)

**Comment by uotalkie** [comment]
By the way, I'm afraid that the code length of wrapper is getting longer and longer..

**Comment by machiavel23** [comment]
I doubt the length is the problem honestly

**Comment by machiavel23** [comment]
so the variables of the wrapper are not passed to the called colab. I expected that. It'll still be enough to test however. Gonna share in the thread as a 4th test file

**Comment by uotalkie** [comment]
![is_wrapped](https://user-images.githubusercontent.com/59263390/71557604-b133bc80-2a8b-11ea-978d-f7bce3238918.JPG)

Variables are passed from wrapper to called, aren't they?

Edit: I didn't know but global variables are always alive unless a VM stops, so the first code above
will give "is_wrapped = true" if the second code is executed. So it's better to delete is_wrapped
when the code are finished executing. like this:
![is_wrapped 2](https://user-images.githubusercontent.com/59263390/71557712-497e7100-2a8d-11ea-97ec-d9482e3e1b4e.JPG)

**Comment by uotalkie** [comment]
Here's yet another revised wrapper to fix is_wrapped above:

[wrapper.zip](https://github.com/Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab/files/4008090/wrapper.zip)

**Comment by machiavel23** [comment]
actually, what you built is a caller more than a wrapper. it will basically ask a cell in a ipynb to brute run. there are only 2 ways to give the variables to the called file with colab. the first one, is to make the caller write the vars in a file, and make the called read the vars from this file, and the other is to make the caller directly write the vars in the called before calling it.

each cell in colab is a subprocess of the main kernel, and global vars are tied to their process, meaning it already need a trick to pass a var from a cell to another (check the code snippets in the table of contents), so imagine to pass a var from a notebook to another...

**Comment by uotalkie** [comment]
First of all, I'm sorry because I feel that my comment is getting out of topic.
If anyone think I'm going too far, please say so.

We can say it a caller or a wrapper. I think name is not so important. 
Also I use term wrappee as the code/notebook executed by run_cell().

I agree with @machiavel23 about passing variables, but that is the case when
multiple VM is running. For example, open two notebooks and execute codes
simultaneously. In this case of wrapper, I think ip.run_cell() is working similarly
to eval() function in javascript. Only one VM is used, and the VM holds every
local and global variables. When VM meets ip.run_cell(), it just executes the code
in run_cell's argument as if the code was there from the beginning. All local and
global variables which VM holds apply to the code in run_cell's argument.

Here's (you might be tired with this, but) my test notes.
[wrapper_test.zip](https://github.com/Kornilov1/AI-Dungeon-2-Anons-modded-Versions-Colab/files/4008563/wrapper_test.zip)
It includes modified version of wrapper (doesn't do git, just read modified version of wrappee)
and Open_CYOAI_Colab_1.22.3_BETA_test.ipynb, a modified version of Open_(omitted)_BETA. 
I added some codes to override default value if the code is wrapped, like this:
![ocyoai_test](https://user-images.githubusercontent.com/59263390/71562941-2d012980-2acb-11ea-81fc-1f38ccaa66c7.JPG)

When I use Open_(omitted)_test.ipynb alone and set AIDungeon / 5000, it works as 
AIDungeon / 5000. When I use wrapper_test.ipynb and set Clover-Edition / 10000,
it works as Clover-Edition / 10000.

Therefore, I think that I can say variables are passed from wrapper to wrappee (or caller to
called) because one VM (or process) is executing the code of wrapper and wrappee.

**Comment by VBPXKSMI** [comment]
@machiavel23 Just letting you know that the general concensus right now is that the disconnect bug seens to be Google fuckery, and that the code is not really at fault here.

While there are a ton of questions still, I think it's a good guess. I had seen reports of the Beta running smothly to some, so maybe the time for merging is near, but you are free to tell me when.
Also, I'm still not so sure if the v1.0 is actually that stable or if it's just confirmation bias / placebo effect, so honestly I'm puzzled in that regard due to the lack of feedback.

**Comment by machiavel23** [comment]
All-in-one or multi : 
Edit: link removed
**If multi-cell wins**:
Edit: link removed
GDrive handling
Edit: link removed

@Kornilov1 Here. drop that in the thread (I can't because of my ban), I'll do a final beta depending on the results, and once this beta will be done, AND once the blanks are filled in the disclaimer OR the phrases removed if we CAN'T fill some blanks, we'll push it as master, make sure everything is sfine and do the release. Looks like a good way to go, no?

EDIT: If you know a better way to gather the opinions of these retards than a strawpoll, feel free to do it like you want. strawpoll is the only one I know. also, it does seem to be google fuckery https://cloud.google.com/compute/docs/instances/preemptible

**Comment by machiavel23** [comment]
Actually, I managed to finally find a working IP, so I send the strawpolls @Kornilov1

**Comment by machiavel23** [comment]
https://github.com/VBPXKSMI/Open-CYOAI-Project/issues/155#issuecomment-572535277

**Comment by machiavel23** [comment]
@uotalkie If you feel the wings of development grow in your back, i suggest you to seriously consider this

https://github.com/voila-dashboards/voila

It might come in handy to provide a good-looking wrapper for the ipynb, it's even officially recommended on the jupyter website

**Comment by uotalkie** [comment]
@machiavel23  Thank you for the information, but I am not interested in making AI Dungeon into standalone web application, which voila is for.

If you are talking about how voila executes a code as wrapper, it seems that voila also uses run_cell to  execute a code.

It's like this:
VoilaExecutePreprocessor.executenb -> VoilaExecutePreprocessor.preprocess -> (super)ExecutePreprocessor.preprocess -> (super) Preprocessor.preprocess -> Preprocessor.preprocess_cell -> (override)ExecutePreprocessor.preprocess_cell -> self.run_cell

Anyway, as I mentioned above, my "wrapping" method was for solving the disconnection issue, and I actually have little interest in python or notebook without AI Dungeon. I'm sorry I could't be any help to solve the issue.

One more thing, about your comment #155, I tried similar method when I was playing AI Dungeon, but disconnection happened. I doubt the method solves the issue.

I used https://bryan7.tistory.com/1077 , for your information. I think this method works only for unattended jobs.

**Comment by machiavel23** [comment]
I'm actually seriously considering converting everything in different .py files and just call them from the cell. the code becomes quite big, and with the widgets, it'll be pretty easy to encapsulate everything in one cell that will call the py associated with the query. with such a light code inside the cell, it could potentially help the disconnect, but only if the code length is indeed at fault. thoughts @VBPXKSMI ?

**Comment by VBPXKSMI** [comment]
@machiavel23 Sorry for reading this so late, was too occupied with the Wiki to bother reading the issues.

I think it could be worth a shot, to see if it helps with the issue or not. I was planning on trying to port things to the Old Colab when I finished the "Starting Page" anyway, but if this resolved the issue it would be far better.
Although I'm not sure if it will help, because it would only help if the lenght of the code is what causes the bug, but if the issue lies on the cell doing too many things at once... I'm not sure it would work. But I'm just being pessimistic here so by all means I think it is still worth trying.
What I know for certain is that you should focus on it only after finishing with the customizator. Working on different things at the same time is kind of difficult (as you can see by my late reply).

**Comment by machiavel23** [comment]
OK, here's my plan.

- Finish the customizator
- adapt the all in one cell to read customizations values from the ini file
- rework each cell to use widgets instead of cells forms to get variables, in order to
- convert each cell to a .py file in order to
- rebuild a new all-in-one cell with only a few tabs and buttons to call the different .py files

gonna take quite some work, I'm not sure I'll have finish before you finish the wiki, so in the case you finish the wiki before what will be colab v2, try to fill the blanks in the disclaimer as much as possible, or delete things where no proof can be found. Also, I was considering the idea to move the disclaimer to the wiki maybe.
